crates/rattler_installs_packages/src/index/package_database.rs

```diff
@@ -472,6 +485,54 @@ impl PackageDb {
         Ok(None)
     }
 
+    async fn get_metadata_stree<'a, A: Borrow<ArtifactInfo>>(
+        &self,
+        artifacts: &'a [A],
+        wheel_builder: &WheelBuilder,
+    ) -> miette::Result<Option<(&'a A, WheelCoreMetadata)>> {
+        let stree = artifacts
+            .iter()
+            .filter(|artifact_info| (*artifact_info).borrow().is::<STree>());
+
+        // Keep track of errors
+        // only print these if we have not been able to find any metadata
+        let mut errors = Vec::new();
+        for ai in stree {
+            let artifact_info: &ArtifactInfo = ai.borrow();
+            let stree_name = artifact_info.filename.as_stree().unwrap_or_else(|| {
+                panic!(
+                    "the specified artifact '{}' does not refer to type requested to read",
+                    artifact_info.filename
+                )
+            });
+            let response = self
+                .get_artifact_and_metadata_by_direct_url(
+                    stree_name.distribution.clone(),
+                    artifact_info.url.clone(),
+                    wheel_builder,
+                )
+                .await;
+
+            match response {
+                Ok(direct_response) => {
+                    let metadata_and_bytes = direct_response.metadata;
+                    self.put_metadata_in_cache(artifact_info, &metadata_and_bytes.0)
+                        .await?;
+                    return Ok(Some((ai, metadata_and_bytes.1)));
+                }
+                Err(err) => {
+                    errors.push(format!(
+                        "error while processing source tree '{}': \n {}",
+                        artifact_info.filename, err
+                    ));
+                    continue;
+                }
+            }
+        }
+
+        Ok(None)
+    }
+
```

Comment:
I think, but I'm not certain that we can generalize this function and `get_metadata_sdists` so we could make a function `get_metadata_source_dependency<T: ArtifactFromSource>()` that takes a Future that should return the (WheelCoreMetadata, Vec<u8>). We can provide a future for.

This part:
```rust

            let response = self
                .get_artifact_and_metadata_by_direct_url(
                    stree_name.distribution.clone(),
                    artifact_info.url.clone(),
                    wheel_builder,
                )
                .await;
```

and in the sdist case for:

```
            let response = self
                .get_artifact_and_metadata_by_direct_url(
                    stree_name.distribution.clone(),
                    artifact_info.url.clone(),
                    wheel_builder,
                )
                .await;
```

this part, then we can re-use the rest of the code.

Reply:
Give it a try, not sure that it will work out :)